cupy/linalg/norms.py

```diff
@@ -268,13 +272,17 @@ def slogdet(a):
 
     diag = cupy.diagonal(lu, axis1=-2, axis2=-1)
 
+    logdet = cupy.log(cupy.abs(diag)).sum(axis=-1)
+
     # ipiv is 1-origin
-    non_zero = (cupy.count_nonzero(ipiv != cupy.arange(1, n + 1), axis=-1) +
-                cupy.count_nonzero(diag < 0, axis=-1))
+    non_zero = cupy.count_nonzero(ipiv != cupy.arange(1, n + 1), axis=-1)
+    if dtype.kind == "f":
+        non_zero += cupy.count_nonzero(diag < 0, axis=-1)
 
     # Note: sign == -1 ** (non_zero % 2)
     sign = (non_zero % 2) * -2 + 1
-    logdet = cupy.log(abs(diag)).sum(axis=-1)
+    if dtype.kind == "c":
+        sign = sign * cupy.prod(diag, axis=-1) / cupy.exp(logdet)
```

Comment:
`cupy.prod(diag, axis=-1)` here might produce several numerical errors for large matrices resulting in inf or error. I wonder if just `cupy.prod(cupy.sign(diag), axis=-1))` is better and allows you to get rid of the `cupy.exp` division.